### ISSUE TITLE
Incorrect constant evaluation warning on correct code

### ISSUE THREAD

**Issue by mooman219** [rust-lang/rust#46553]
It looks like I'm incorrectly being warned of a constant evaluation error on valid code. The code runs just fine and is not violating the constant function requirements.

Offending code:
```rust
#![feature(const_fn)]

pub struct Data<T> {
    function: fn() -> T,
}

impl<T> Data<T> {
    pub const fn new(function: fn() -> T) -> Data<T> {
        Data {
            function: function,
        }
    }
}

pub static DATA: Data<i32> = Data::new(|| {
    413i32
});

fn main() {
    print!("{:?}", (DATA.function)());
}
```
Incorrect error:
```rust
warning: constant evaluation error: unsupported constant expr
  --> src/main.rs:15:30
   |
15 |   pub static DATA: Data<i32> = Data::new(|| {
   |  ______________________________^
16 | |     413i32
17 | | });
   | |__^
   |
   = note: #[warn(const_err)] on by default
```
[Example of the issue on the playground in nightly](https://play.rust-lang.org/?gist=201319d1da5c170986d22d096d493f0d&version=nightly).

Ping #24111

**Comment by oli-obk** [comment]
This looks very much like the AST const evaluator simply doesn't know about closures under some conditions. I'd wager it has to do with the generics.

I suggest papering over the issue by placing `unsupported constant expr` into the list of ignored const eval errors. This will simply work once we get miri, so there's no point in putting more effort into it.

**Comment by AgustinCB** [comment]
I'd be happy to take the task. Where such a list of const ignored errors?

**Comment by oli-obk** [comment]
@AgustinCB great!

The list is

https://github.com/rust-lang/rust/blob/24840dab0bef0047beb428e1c15e9a851f732dcc/src/librustc_passes/consts.rs#L134-L138

You probably just need to ignore 

https://github.com/rust-lang/rust/blob/2f84fb5cc1ac5f04bdf5a281997e02bcf4fc18d9/src/librustc/middle/const_val.rs#L164

And fixup any tests that were affected
